cli/src/lib.rs

```diff
@@ -529,71 +506,20 @@ macro_rules! mutate_nested_option {
 /// # Errors
 /// - If provided user configuration is invalid or incomplete
 /// - If genesis config is invalid
-pub fn read_config(
-    path: &Path,
+pub fn read_config_and_genesis(
+    path: impl AsRef<Path>,
     submit_genesis: bool,
-) -> Result<(Configuration, Option<GenesisNetwork>)> {
-    let config = ConfigurationProxy::default();
-
-    let config = if let Some(actual_config_path) = path
-        .try_resolve()
-        .wrap_err("Failed to resolve configuration file")?
-    {
-        let mut cfg = config.override_with(ConfigurationProxy::from_path(&*actual_config_path));
-        let config_dir = actual_config_path
-            .parent()
-            .expect("If config file was read, than it should have a parent. It is a bug.");
-
-        // careful here: `genesis.file` might be a path relative to the config file.
-        // we need to resolve it before proceeding
-        // TODO: move this logic into `iroha_config`
-        //       https://github.com/hyperledger/iroha/issues/4161
-        let join_to_config_dir = |x: &mut PathBuf| {
-            *x = config_dir.join(&x);
-        };
-        mutate_nested_option!(cfg, [genesis, file, self], join_to_config_dir);
-        mutate_nested_option!(cfg, [snapshot, dir_path], join_to_config_dir);
-        mutate_nested_option!(cfg, [kura, block_store_path], join_to_config_dir);
-        mutate_nested_option!(cfg, [telemetry, file, self], join_to_config_dir);
-
-        cfg
-    } else {
-        config
-    };
+) -> Result<(Config, Option<GenesisNetwork>)> {
+    use iroha_config::parameters::actual::Genesis;
 
-    // it is not chained to the previous expressions so that config proxy from env is evaluated
-    // after reading a file
-    let config = config.override_with(
-        ConfigurationProxy::from_std_env().wrap_err("Failed to build configuration from env")?,
-    );
-
-    let config = config
-        .build()
-        .wrap_err("Failed to finalize configuration")?;
+    let config = Config::load(path, CliContext { submit_genesis })
+        .wrap_err("failed to load configuration")?;
```

Comment:
after #4225 it will be ok for any peer to submit genesis. You may apply that change already and remove `CliContext`

Reply:
Oh, would be nice if this issue is resolved before the merge of this PR. I will remove `CliContext` upon rebasing on top of #4225 fix

Reply:
sadly, that's hardly going to happen. It's more likely that #4255 will come after this PR

Reply:
Actually not that sad, because it might be painful to rebase this huge PR.